documentcloud/documents/models.py

```diff
@@ -810,6 +833,46 @@ def __str__(self):
         return self.title
 
 
+class Revision(models.Model):
+    """A saved version of the document made before it was edited"""
+
+    document = models.ForeignKey(
+        verbose_name=_("document"),
+        to="documents.Document",
+        on_delete=models.CASCADE,
+        related_name="revisions",
+    )
+    user = models.ForeignKey(
+        verbose_name=_("user"),
+        to="users.User",
+        on_delete=models.PROTECT,
+        related_name="revisions",
+    )
+    created_at = AutoCreatedField(
+        _("created at"),
+    )
+    version = models.PositiveIntegerField(
+        _("version"),
+        default=0,
+    )
+    comment = models.CharField(
+        _("comment"),
+        max_length=255,
+    )
+
+    class Meta:
+        unique_together = [("document", "version")]
+        ordering = ("version",)
+
+    def copy(self):
+        """Copy the current PDF to this revision"""
+        destination = path.doc_revision_path(
+            self.document.pk, self.document.slug, self.version
+        )
+        if not storage.exists(destination):
+            storage.copy(self.document.doc_path, destination)
```

Comment:
Hopefully, this never comes up, but since we're checking to see if something already exists for this version before copying, should we log/alert ourselves if something does already exist?

Reply:
I had a scenario in mind where this could happen in normal operation, but now cannot think of it.  But I will add a log here so we can be aware if it happens.